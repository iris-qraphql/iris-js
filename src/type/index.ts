export type { Path as ResponsePath } from '../jsutils/Path';

export {
  // Predicate
  isSchema,
  // Assertion
  assertSchema,
  // GraphQL Schema definition
  GraphQLSchema,
} from './schema';
export type { GraphQLSchemaConfig, GraphQLSchemaExtensions } from './schema';

export type { GraphQLScalarType } from './definition';
export {
  isType,
  isScalarType,
  isObjectType,
  isUnionType,
  isEnumType,
  isInputObjectType,
  isListType,
  isNonNullType,
  isInputType,
  isOutputType,
  isLeafType,
  isResolverType,
  isAbstractType,
  isWrappingType,
  isNullableType,
  isNamedType,
  isRequiredArgument,
  isRequiredInputField,
  assertScalarType,
  assertObjectType,
  assertResolverType,
  assertListType,
  assertNonNullType,
  assertLeafType,
  assertCompositeType,
  assertAbstractType,
  getNullableType,
  getNamedType,
  GraphQLList,
  GraphQLNonNull,
} from './definition';

export type {
  GraphQLType,
  GraphQLInputType,
  GraphQLOutputType,
  GraphQLLeafType,
  IrisResolverType,
  GraphQLWrappingType,
  GraphQLNullableType,
  GraphQLNamedType,
  GraphQLNamedInputType,
  GraphQLNamedOutputType,
  ThunkObjMap,
  GraphQLArgument,
  GraphQLArgumentConfig,
  IrisDataVariant,
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLFieldConfigArgumentMap,
  GraphQLFieldConfigMap,
  GraphQLFieldMap,
  GraphQLFieldResolver,
  GraphQLInputField,
  GraphQLIsTypeOfFn,
  GraphQLResolveInfo,
  GraphQLTypeResolver,
  GraphQLScalarSerializer,
  GraphQLScalarValueParser,
  GraphQLScalarLiteralParser,
} from './definition';

export {
  // Predicate
  isDirective,
  // Assertion
  assertDirective,
  // Directives Definition
  GraphQLDirective,
  // Built-in Directives defined by the Spec
  isSpecifiedDirective,
  specifiedDirectives,
  GraphQLIncludeDirective,
  GraphQLSkipDirective,
  GraphQLDeprecatedDirective,
  GraphQLSpecifiedByDirective,
  // Constant Deprecation Reason
  DEFAULT_DEPRECATION_REASON,
} from './directives';

export type {
  GraphQLDirectiveConfig,
  GraphQLDirectiveExtensions,
} from './directives';

// Common built-in scalar instances.
export {
  // Predicate
  isSpecifiedScalarType,
  // Standard GraphQL Scalars
  specifiedScalarTypes,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLBoolean,
  GraphQLID,
  // Int boundaries constants
  GRAPHQL_MAX_INT,
  GRAPHQL_MIN_INT,
} from './scalars';

export {
  // Predicate
  isIntrospectionType,
  // GraphQL Types for introspection.
  introspectionTypes,
  __Schema,
  __Directive,
  __DirectiveLocation,
  __Type,
  __Field,
  __InputValue,
  __EnumValue,
  __TypeKind,
  // "Enum" of Type Kinds
  TypeKind,
  // Meta-field definitions.
  SchemaMetaFieldDef,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
} from './introspection';

// Validate GraphQL schema.
export { validateSchema, assertValidSchema } from './validate';

// Upholds the spec rules about naming.
export { assertName, assertEnumValueName } from './assertName';
