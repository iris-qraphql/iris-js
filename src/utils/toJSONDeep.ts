import { isObjectLike } from '../jsutils/isObjectLike';
import { mapValue } from '../jsutils/mapValue';

export function toJSONDeep(value: unknown): unknown {
  if (!isObjectLike(value)) {
    return value;
  }

  if (typeof value.toJSON === 'function') {
    return value.toJSON();
  }

  if (Array.isArray(value)) {
    return value.map(toJSONDeep);
  }

  return mapValue(value, toJSONDeep);
}

export function toJSONError(fn: () => unknown) {
  try {
    fn();
    return undefined;
  } catch (error) {
    return toJSONDeep(error);
  }
}
