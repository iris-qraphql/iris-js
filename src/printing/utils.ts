import type { Maybe } from '../utils/type-level';

/**
 * Given maybeArray, print an empty string if it is null or empty, otherwise
 * print all items together separated by separator if provided
 */
export function join(
  maybeArray: Maybe<ReadonlyArray<string | undefined>>,
  separator = '',
): string {
  return maybeArray?.filter((x) => x).join(separator) ?? '';
}

/**
 * Given array, print each item on its own line, wrapped in an indented `{ }` block.
 */
export const block = (array: ReadonlyArray<string | undefined>) =>
  array.length > 0 ? `{\n${indent(join(array, '\n'))}\n}` : '{}';

/**
 * If maybeString is not null or empty, then wrap with start and end, otherwise print an empty string.
 */
export function wrap(
  start: string,
  maybeString: Maybe<string>,
  end: string = '',
): string {
  return maybeString != null && maybeString !== ''
    ? start + maybeString + end
    : '';
}

export function indent(str: string): string {
  return wrap('  ', str.replace(/\n/g, '\n  '));
}
