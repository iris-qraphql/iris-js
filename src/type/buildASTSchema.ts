import type { ParseOptions, Source } from 'graphql';

import type { ObjMap } from '../jsutils/ObjMap';
import { keyMap } from '../jsutils/ObjMap';

import type {
  ArgumentDefinitionNode,
  DirectiveDefinitionNode,
  DocumentNode,
  FieldDefinitionNode,
  NamedTypeNode,
  ResolverVariantDefinitionNode,
  Role,
  TypeDefinitionNode,
  TypeNode,
  VariantDefinitionNode,
} from '../language/ast';
import { IrisKind } from '../language/kinds';
import { parse } from '../language/parser';
import { isTypeDefinitionNode } from '../language/predicates';

import { validateSDL } from '../validation/validate';

import { valueFromAST } from '../conversion/valueFromAST';
import { getDirectiveValues } from '../conversion/values';
import type { ConfigMap, Maybe } from '../utils/type-level';

import type {
  IrisArgument,
  IrisFieldConfig,
  IrisNamedType,
  IrisResolverVariantConfig,
  IrisType,
  IrisVariant,
} from './definition';
import { IrisDataType, IrisResolverType, IrisTypeRef } from './definition';
import { GraphQLDeprecatedDirective, GraphQLDirective } from './directives';
import { specifiedScalarTypes } from './scalars';
import type { IrisSchemaValidationOptions } from './schema';
import { IrisSchema } from './schema';

export function buildSchema(
  source: string | Source,
  options?: IrisSchemaValidationOptions & ParseOptions,
): IrisSchema {
  const document = parse(source, { noLocation: options?.noLocation });

  return buildASTSchema(document, { ...options });
}

export function buildASTSchema(
  documentAST: DocumentNode,
  options?: IrisSchemaValidationOptions,
): IrisSchema {
  if (options?.assumeValid !== true && options?.assumeValidSDL !== true) {
    const errors = validateSDL(documentAST);
    if (errors.length !== 0) {
      throw new Error(errors.map((error) => error.message).join('\n\n'));
    }
  }

  const directiveDefs: Array<DirectiveDefinitionNode> = [];
  const typeMap: Record<string, IrisNamedType> = {};

  documentAST.definitions.forEach((def) => {
    if (isTypeDefinitionNode(def)) {
      const name = def.name.value;
      typeMap[name] = stdTypeMap[name] ?? buildType(def);
    } else if (def.kind === IrisKind.DIRECTIVE_DEFINITION) {
      directiveDefs.push(def);
    }
  });

  return new IrisSchema({
    description: undefined,
    types: Object.values(typeMap),
    query: typeMap.Query as IrisResolverType,
    mutation: typeMap.Mutation as IrisResolverType,
    subscription: typeMap.Subscription as IrisResolverType,
    directives: directiveDefs.map(buildDirective),
    assumeValid: options?.assumeValid ?? false,
  });

  function getNamedType(node: ResolverVariantDefinitionNode): IrisResolverType;
  function getNamedType(node: NamedTypeNode): IrisNamedType;
  function getNamedType(
    node: NamedTypeNode | ResolverVariantDefinitionNode,
  ): IrisNamedType {
    const name = node.name.value;
    const type = stdTypeMap[name] ?? typeMap[name];

    if (type === undefined) {
      throw new Error(`Unknown type: "${name}".`);
    }
    return type;
  }

  function getWrappedType(node: TypeNode): IrisType {
    if (node.kind === IrisKind.LIST_TYPE) {
      return new IrisTypeRef('LIST', getWrappedType(node.type));
    }
    if (node.kind === IrisKind.MAYBE_TYPE) {
      return new IrisTypeRef('MAYBE', getWrappedType(node.type));
    }
    return getNamedType(node);
  }

  function buildDirective(node: DirectiveDefinitionNode): GraphQLDirective {
    return new GraphQLDirective({
      name: node.name.value,
      description: node.description?.value,
      // @ts-expect-error
      locations: node.locations.map(({ value }) => value),
      isRepeatable: node.repeatable,
      args: buildArgumentMap(node.arguments),
      astNode: node,
    });
  }

  function buildArgumentMap(
    args: Maybe<ReadonlyArray<ArgumentDefinitionNode>>,
  ): ConfigMap<IrisArgument> {
    const argsNodes = /* c8 ignore next */ args ?? [];

    const argConfigMap = Object.create(null);
    for (const arg of argsNodes) {
      // Note: While this could make assertions to get the correctly typed
      // value, that would throw immediately while type system validation
      // with validateSchema() will produce more actionable results.
      const type: any = getWrappedType(arg.type);

      argConfigMap[arg.name.value] = {
        type,
        description: arg.description?.value,
        defaultValue: valueFromAST(arg.defaultValue, type),
        deprecationReason: getDeprecationReason(arg),
        astNode: arg,
      };
    }
    return argConfigMap;
  }

  function buildFieldMap<R extends Role>(
    fields: ReadonlyArray<FieldDefinitionNode>,
  ): ObjMap<IrisFieldConfig<R>> {
    const entries = fields.map((field) => {
      const type: any = getWrappedType(field.type);
      return [
        field.name.value,
        {
          type,
          description: field.description?.value,
          deprecationReason: getDeprecationReason(field),
          astNode: field,
          args: field.arguments ? buildArgumentMap(field.arguments) : undefined,
        },
      ];
    });

    return Object.fromEntries(entries);
  }

  function buildVariant<R extends Role>(
    role: R,
    astNode: ResolverVariantDefinitionNode,
  ): IrisResolverVariantConfig | IrisVariant<'data'> {
    const name = astNode.name.value;
    const description = astNode.description?.value;
    const deprecationReason = getDeprecationReason(astNode);

    if (!astNode.fields) {
      return {
        name,
        description,
        deprecationReason,
        astNode,
        type: getNamedType(astNode),
      };
    }

    if (role === 'data') {
      return {
        name,
        description,
        deprecationReason,
        fields: astNode.fields
          ? () => buildFieldMap<'data'>(astNode?.fields ?? [])
          : undefined,
        astNode,
      } as IrisVariant<'data'>;
    }

    return {
      name,
      description,
      deprecationReason,
      fields: buildFieldMap<'resolver'>(astNode.fields ?? []),
      astNode,
    };
  }

  function buildType(astNode: TypeDefinitionNode): IrisNamedType {
    const name = astNode.name.value;

    switch (astNode.kind) {
      case IrisKind.RESOLVER_TYPE_DEFINITION: {
        return new IrisResolverType({
          name,
          description: astNode.description?.value,
          variants: () =>
            astNode.variants.map((v) => buildVariant('resolver', v)),
          astNode,
        });
      }
      case IrisKind.DATA_TYPE_DEFINITION: {
        return new IrisDataType({
          name,
          description: astNode.description?.value,
          variants: astNode.variants.map((v) => buildVariant('data', v)) as any,
          astNode,
        });
      }
    }
  }
}

const stdTypeMap = keyMap([...specifiedScalarTypes], (type) => type.name);

const getDeprecationReason = (
  node: FieldDefinitionNode | ArgumentDefinitionNode | VariantDefinitionNode,
): Maybe<string> =>
  getDirectiveValues(GraphQLDeprecatedDirective, node)?.reason as string;
