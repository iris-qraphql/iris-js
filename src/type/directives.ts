import { assertName } from 'graphql';

import type { Maybe } from '../jsutils/Maybe';

import type { DirectiveDefinitionNode } from '../language/ast';
import { DirectiveLocation } from '../language/directiveLocation';

import { inspect, instanceOf } from '../utils/legacy';
import type { ConfigMap } from '../utils/type-level';

import type { IrisArgument } from './definition';
import { buildArguments } from './definition';
import { IrisString } from './scalars';

/**
 * Test if the given value is a GraphQL directive.
 */
export function isDirective(directive: unknown): directive is GraphQLDirective {
  return instanceOf(directive, GraphQLDirective);
}

export function assertDirective(directive: unknown): GraphQLDirective {
  if (!isDirective(directive)) {
    throw new Error(
      `Expected ${inspect(directive)} to be a GraphQL directive.`,
    );
  }
  return directive;
}

/**
 * Custom extensions
 *
 * @remarks
 * Use a unique identifier name for your extension, for example the name of
 * your library or project. Do not use a shortened identifier as this increases
 * the risk of conflicts. We recommend you add at most one extension field,
 * an object which can contain all the values you need.
 */
export type GraphQLDirectiveExtensions = Record<string, unknown>;

/**
 * Directives are used by the GraphQL runtime as a way of modifying execution
 * behavior. Type system creators will usually not create these directly.
 */
export class GraphQLDirective {
  name: string;
  description: Maybe<string>;
  locations: ReadonlyArray<DirectiveLocation>;
  args: ReadonlyArray<IrisArgument>;
  isRepeatable: boolean;
  astNode: Maybe<DirectiveDefinitionNode>;

  constructor(config: Readonly<GraphQLDirectiveConfig>) {
    this.name = assertName(config.name);
    this.description = config.description;
    this.locations = config.locations;
    this.isRepeatable = config.isRepeatable ?? false;
    this.astNode = config.astNode;
    this.args = buildArguments(config.args ?? {});
  }

  get [Symbol.toStringTag]() {
    return 'GraphQLDirective';
  }

  toString(): string {
    return '@' + this.name;
  }

  toJSON(): string {
    return this.toString();
  }
}

export interface GraphQLDirectiveConfig {
  name: string;
  description?: Maybe<string>;
  locations: ReadonlyArray<DirectiveLocation>;
  args?: ConfigMap<IrisArgument>;
  isRepeatable?: Maybe<boolean>;
  astNode?: Maybe<DirectiveDefinitionNode>;
}

/**
 * Used to declare element of a GraphQL schema as deprecated.
 */
export const GraphQLDeprecatedDirective: GraphQLDirective =
  new GraphQLDirective({
    name: 'deprecated',
    description: 'Marks an element of a GraphQL schema as no longer supported.',
    locations: [
      DirectiveLocation.FIELD_DEFINITION,
      DirectiveLocation.ARGUMENT_DEFINITION,
      DirectiveLocation.INPUT_FIELD_DEFINITION,
      DirectiveLocation.ENUM_VALUE,
    ],
    args: {
      reason: {
        type: IrisString,
        description:
          'Explains why this element was deprecated, usually also including a suggestion for how to access supported similar data. Formatted using the Markdown syntax, as specified by [CommonMark](https://commonmark.org/).',
        defaultValue: '',
      },
    },
  });

/**
 * The full list of specified directives.
 */
export const specifiedDirectives: ReadonlyArray<GraphQLDirective> =
  Object.freeze([GraphQLDeprecatedDirective]);

export function isSpecifiedDirective(directive: GraphQLDirective): boolean {
  return specifiedDirectives.some(({ name }) => name === directive.name);
}
