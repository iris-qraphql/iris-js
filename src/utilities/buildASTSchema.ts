import type { ParseOptions, Source } from 'graphql';

import type { Maybe } from '../jsutils/Maybe';
import type { ObjMap } from '../jsutils/ObjMap';
import { keyMap } from '../jsutils/ObjMap';

import type {
  ArgumentDefinitionNode,
  DataFieldDefinitionNode,
  DirectiveDefinitionNode,
  DocumentNode,
  FieldDefinitionNode,
  NamedTypeNode,
  ResolverVariantDefinitionNode,
  TypeDefinitionNode,
  TypeNode,
  VariantDefinitionNode,
} from '../language/ast';
import { IrisKind } from '../language/kinds';
import { parse } from '../language/parser';
import { isTypeDefinitionNode } from '../language/predicates';

import type {
  GraphQLArgument,
  GraphQLFieldConfig,
  IrisDataVariantConfig,
  IrisDataVariantField,
  IrisNamedType,
  IrisResolverVariantConfig,
  IrisType,
} from '../type/definition';
import {
  IrisDataType,
  IrisResolverType,
  IrisTypeRef,
} from '../type/definition';
import {
  GraphQLDeprecatedDirective,
  GraphQLDirective,
} from '../type/directives';
import { specifiedScalarTypes } from '../type/scalars';
import type { IrisSchemaValidationOptions } from '../type/schema';
import { IrisSchema } from '../type/schema';

import { validateSDL } from '../validation/validate';

import type { ConfigMap } from '../utils/type-level';

import { valueFromAST } from './valueFromAST';
import { getDirectiveValues } from './values';

export function buildSchema(
  source: string | Source,
  options?: IrisSchemaValidationOptions & ParseOptions,
): IrisSchema {
  const document = parse(source, { noLocation: options?.noLocation });

  return buildASTSchema(document, { ...options });
}

export function buildASTSchema(
  documentAST: DocumentNode,
  options?: IrisSchemaValidationOptions,
): IrisSchema {
  if (options?.assumeValid !== true && options?.assumeValidSDL !== true) {
    const errors = validateSDL(documentAST);
    if (errors.length !== 0) {
      throw new Error(errors.map((error) => error.message).join('\n\n'));
    }
  }

  const directiveDefs: Array<DirectiveDefinitionNode> = [];
  const typeMap: Record<string, IrisNamedType> = {};

  documentAST.definitions.forEach((def) => {
    if (isTypeDefinitionNode(def)) {
      const name = def.name.value;
      typeMap[name] = stdTypeMap[name] ?? buildType(def);
    } else if (def.kind === IrisKind.DIRECTIVE_DEFINITION) {
      directiveDefs.push(def);
    }
  });

  return new IrisSchema({
    description: undefined,
    types: Object.values(typeMap),
    query: typeMap.Query as IrisResolverType,
    mutation: typeMap.Mutation as IrisResolverType,
    subscription: typeMap.Subscription as IrisResolverType,
    directives: directiveDefs.map(buildDirective),
    assumeValid: options?.assumeValid ?? false,
  });

  function getNamedType(node: NamedTypeNode): IrisNamedType {
    const name = node.name.value;
    const type = stdTypeMap[name] ?? typeMap[name];

    if (type === undefined) {
      throw new Error(`Unknown type: "${name}".`);
    }
    return type;
  }

  function getWrappedType(node: TypeNode): IrisType {
    if (node.kind === IrisKind.LIST_TYPE) {
      return new IrisTypeRef('LIST', getWrappedType(node.type));
    }
    if (node.kind === IrisKind.MAYBE_TYPE) {
      return new IrisTypeRef('MAYBE', getWrappedType(node.type));
    }
    return getNamedType(node);
  }

  function buildDirective(node: DirectiveDefinitionNode): GraphQLDirective {
    return new GraphQLDirective({
      name: node.name.value,
      description: node.description?.value,
      // @ts-expect-error
      locations: node.locations.map(({ value }) => value),
      isRepeatable: node.repeatable,
      args: buildArgumentMap(node.arguments),
      astNode: node,
    });
  }

  function buildArgumentMap(
    args: Maybe<ReadonlyArray<ArgumentDefinitionNode>>,
  ): ConfigMap<GraphQLArgument> {
    const argsNodes = /* c8 ignore next */ args ?? [];

    const argConfigMap = Object.create(null);
    for (const arg of argsNodes) {
      // Note: While this could make assertions to get the correctly typed
      // value, that would throw immediately while type system validation
      // with validateSchema() will produce more actionable results.
      const type: any = getWrappedType(arg.type);

      argConfigMap[arg.name.value] = {
        type,
        description: arg.description?.value,
        defaultValue: valueFromAST(arg.defaultValue, type),
        deprecationReason: getDeprecationReason(arg),
        astNode: arg,
      };
    }
    return argConfigMap;
  }

  function buildFieldMap(
    node: ReadonlyArray<FieldDefinitionNode>,
  ): ObjMap<GraphQLFieldConfig<unknown, unknown>>;
  function buildFieldMap(
    fields: ReadonlyArray<DataFieldDefinitionNode>,
  ): ObjMap<IrisDataVariantField>;
  function buildFieldMap(
    fields: ReadonlyArray<FieldDefinitionNode>,
  ): ObjMap<GraphQLFieldConfig<unknown, unknown>> {
    const entries = fields.map((field) => {
      const type: any = getWrappedType(field.type);
      return [
        field.name.value,
        {
          type,
          description: field.description?.value,
          deprecationReason: getDeprecationReason(field),
          astNode: field,
          args: field.arguments ? buildArgumentMap(field.arguments) : undefined,
        },
      ];
    });

    return Object.fromEntries(entries);
  }

  function resolveDataVariant(
    value: VariantDefinitionNode,
  ): IrisDataVariantConfig {
    return {
      name: value.name.value,
      description: value.description?.value,
      deprecationReason: getDeprecationReason(value),
      astNode: value,
      // @ts-expect-error
      fields: value.fields
        ? () => buildFieldMap(value?.fields ?? [])
        : undefined,
    };
  }

  function buildResolverVariant(
    variantNode: ResolverVariantDefinitionNode,
  ): IrisResolverVariantConfig {
    const name = variantNode.name.value;
    const description = variantNode.description?.value;

    if (variantNode.fields) {
      return {
        name,
        description,
        fields: () => buildFieldMap(variantNode.fields ?? []),
        astNode: variantNode,
      };
    }

    return {
      name,
      description,
      astNode: variantNode,
      // @ts-expect-error
      type: () => getNamedType(variantNode),
    };
  }

  function buildType(astNode: TypeDefinitionNode): IrisNamedType {
    const name = astNode.name.value;

    switch (astNode.kind) {
      case IrisKind.RESOLVER_TYPE_DEFINITION: {
        return new IrisResolverType({
          name,
          description: astNode.description?.value,
          variants: astNode.variants.map(buildResolverVariant),
          astNode,
        });
      }
      case IrisKind.DATA_TYPE_DEFINITION: {
        return new IrisDataType({
          name,
          description: astNode.description?.value,
          variants: astNode.variants.map(resolveDataVariant),
          astNode,
        });
      }
    }
  }
}

const stdTypeMap = keyMap([...specifiedScalarTypes], (type) => type.name);

const getDeprecationReason = (
  node: FieldDefinitionNode | ArgumentDefinitionNode | VariantDefinitionNode,
): Maybe<string> =>
  getDirectiveValues(GraphQLDeprecatedDirective, node)?.reason as string;
