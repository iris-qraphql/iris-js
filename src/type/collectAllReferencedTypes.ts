import type { Role } from '../language/ast';

import type { IrisType, IrisTypeDefinition, IrisVariant } from './definition';
import { unwrapType } from './definition';
import type { GraphQLDirective } from './directives';
import { isDirective } from './directives';

export const collectAllReferencedTypes = (
  types: ReadonlyArray<IrisTypeDefinition>,
  directives: ReadonlyArray<GraphQLDirective>,
): Set<IrisTypeDefinition> => {
  const allReferencedTypes: Set<IrisTypeDefinition> = new Set(types);

  types.forEach((type) => {
    allReferencedTypes.delete(type);
    collectReferencedTypes(type, allReferencedTypes);
  });

  collectDirectiveTypes(directives, allReferencedTypes);
  return allReferencedTypes;
};

const collectDirectiveTypes = (
  directives: ReadonlyArray<GraphQLDirective>,
  allReferencedTypes: Set<IrisTypeDefinition>,
) => {
  for (const directive of directives) {
    // Directives are not validated until validateSchema() is called.
    if (isDirective(directive)) {
      directive.args.forEach((arg) =>
        collectReferencedTypes(arg.type, allReferencedTypes),
      );
    }
  }
};

function collectReferencedTypes(
  type: IrisType,
  typeSet: Set<IrisTypeDefinition>,
): void {
  const namedType = unwrapType(type);

  if (typeSet.has(namedType)) {
    return;
  }

  typeSet.add(namedType);
  namedType
    .variants()
    .forEach((v) => exploreVariant(v as IrisVariant<'resolver'>, typeSet));
}

const exploreVariant = (
  variant: IrisVariant<Role>,
  typeSet: Set<IrisTypeDefinition>,
) => {
  if (variant.type) {
    collectReferencedTypes(variant.type, typeSet);
  }

  Object.values(variant.fields ?? {}).forEach((field) => {
    collectReferencedTypes(field.type, typeSet);
    if ('args' in field) {
      field.args?.forEach?.((arg) => collectReferencedTypes(arg.type, typeSet));
    }
  });
};
