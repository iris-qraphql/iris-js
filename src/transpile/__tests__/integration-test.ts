import { buildSchema } from '../../type/buildASTSchema';

import { iris } from '../../index';
import { toJSONDeep } from '../../utils/toJSONDeep';

describe('Simple Integration', () => {
  const schema = buildSchema(`
  resolver Query = {
    hello: String
  }
`);

  const rootValue = {
    hello: () => 'world',
  };

  it('hello world App', async () => {
    const result = await iris({ schema, rootValue, source: '{ hello }' });

    expect(toJSONDeep(result)).toEqual({
      data: {
        hello: 'world',
      },
    });
  });
});

describe('Sophisticated Integration', () => {
  const schema = buildSchema(`
  data Lifespan
    = Immortal {}
    | Limited { max: Int? }

  resolver God = {
    name: String
    lifespan: Lifespan
  }

  resolver Deity
    = God
    | Titan { name: String }

  resolver Query = {
    deities(lifespan: Lifespan?): [Deity]
  }
  `);

  const validDeities = () => [
    {
      __typename: 'God',
      name: 'Morpheus',
      lifespan: 'Immortal',
    },
    {
      __typename: 'Deity_Titan',
      name: 'Cronos',
    },
  ];

  const api = (deities: () => Array<unknown>, query: string) =>
    iris({ schema, rootValue: { deities }, source: query });

  it('union type names', async () => {
    const result = await api(validDeities, '{ deities { __typename } }');

    expect(toJSONDeep(result)).toMatchSnapshot();
  });

  it('conditional union selections', async () => {
    const result = await api(
      validDeities,
      `{ 
        deities { 
          __typename
          ... on God {
            name
            lifespan
          }
          ... on Deity_Titan {
            name
          }
        } 
    }`,
    );

    expect(toJSONDeep(result)).toMatchSnapshot();
  });

  it('conditional union selections', async () => {
    const result = await api(
      () => [
        {
          __typename: 'God',
          name: 'Morpheus',
          lifespan: 'Thor',
        },
      ],
      `{ 
        deities { 
          __typename
          ... on God {
            name
            lifespan
          }
          ... on Deity_Titan {
            name
          }
        } 
    }`,
    );

    expect(toJSONDeep(result)).toMatchSnapshot();
  });
});
