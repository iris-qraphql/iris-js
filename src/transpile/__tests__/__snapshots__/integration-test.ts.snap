// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sophisticated Integration conditional union selections 1`] = `
Object {
  "data": Object {
    "deities": Array [
      Object {
        "__typename": "God",
        "lifespan": "Immortal",
        "name": "Morpheus",
      },
      Object {
        "__typename": "Deity_Titan",
        "name": "Cronos",
      },
    ],
  },
}
`;

exports[`Sophisticated Integration conditional union selections 2`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 13,
          "line": 6,
        },
      ],
      "message": "Data \\"Lifespan\\" cannot represent value: \\"Thor\\"",
      "path": Array [
        "deities",
        0,
        "lifespan",
      ],
    },
  ],
}
`;

exports[`Sophisticated Integration union type names 1`] = `
Object {
  "data": Object {
    "deities": Array [
      Object {
        "__typename": "God",
      },
      Object {
        "__typename": "Deity_Titan",
      },
    ],
  },
}
`;
